### ISSUE TITLE
CUDA non-determinism on identical requests

### ISSUE THREAD

**Issue by phiharri** [ggerganov/llama.cpp#2838]
When layers are offloaded with CUDA, sending identical requests to the examples/server completion API returns a different response the "first time":
```
$ for x in `seq 5`; do curl -s -X POST --url 'http://miku:8080/completion' --data '{"prompt":"Some random words:","n_predict":50,"seed":1337}' | jq '.content' ; done
" hollow, glowing, tinkling, crunchy, sinking"
" apartment, blouse, bobby, carousel"
" apartment, blouse, bobby, carousel"
" apartment, blouse, bobby, carousel"
" apartment, blouse, bobby, carousel"
```

This seems cache related as responses then remain the same until a different prompt is processed, after which the differing first response occurs again:
```
$ curl -s -X POST --url 'http://miku:8080/completion' --data '{"prompt":"Building a website is as simple as","n_predict":0}' >/dev/null

$ for x in `seq 5`; do curl -s -X POST --url 'http://miku:8080/completion' --data '{"prompt":"Some random words:","n_predict":50,"seed":1337}' | jq '.content' ; done
" hollow, glowing, tinkling, crunchy, sinking"
" apartment, blouse, bobby, carousel"
" apartment, blouse, bobby, carousel"
..
```

## Expected Behaviour
Output should remain the same when parameters and seed are constant. 

## Other Observations
* Not observed with Metal offload.
* Not observed without CUDA offload (interestingly neither with small n-gpu-layers, eg. CodeLlama-34b shows this behaviour with `-ngl 3` but not with `-ngl 2`).
* Behaviour observed with both non-K and K-quants.
* The first response ("hollow, glowing.." above) is what examples/main returns with the same parameters.

## Environment
* Verified behaviour on latest master commit, compiled with `LLAMA_CUBLAS=1 make -j`
* Linux 5.15.0-79-generic x86_64
* NVIDIA 535.86.05
* CUDA 12.2
* Python 3.10.12
* GNU Make 4.3
* g++ 11.4.0

Thanks for reading! 😎

**Comment by staviq** [comment]
Reproducible on c10704d01e21e3dbe4d6ca1026ebff85349dd239
```
server -m /storage/models/mythomax-l2-13b.ggmlv3.q5_K_M.gguf --port 7860 -c 4096 -b 512 -t 6 -ngl 43 -ts 80,101 --mlock
```

```
for x in `seq 5`; do curl -s -X POST --url 'http://localhost:7860/completion' --data '{"prompt":"Some random words:","n_predict":50,"seed":1337}' | jq '.content' ; done
" business, communication, community, education, health"
" greek, hellenistic, language, papyrus, text"
" greek, hellenistic, language, papyrus, text"
" greek, hellenistic, language, papyrus, text"
" greek, hellenistic, language, papyrus, text"
```

**Comment by ReactorScram** [comment]
Even on CPU inference I can't get any determinism from the example server at all:

```
$ for x in `seq 5`; do curl -s -X POST --url http://localhost:8080/completion --header "Content-Type: application/json" --data '{"prompt": "Cells interlinked, within cells interlinked, within a single stem.", "n_predict": 16, "seed":1337}' | jq '.content' ; done
"\nA network of life, the most basic of forms, intertwined."
"\nA single stem, from which all others come.\nThis is not about"
" The stem is the source of life, the starting point for all. Leaves"
" The stem is the root of all life,\nFrom which the branches extend,"
" The stem is the root of all life,\nFrom which roots grow in multiple"
```

I'll probably try using the library directly, but I'm trying to use it in Rust and 2 of the popular Rust wrappers already had issues where they segfault or act odd, which is strange, because the C++ code itself hasn't segfaulted on me yet.

**Comment by ggerganov** [comment]
We seem to be ignoring the provided seed:

https://github.com/ggerganov/llama.cpp/blob/6bb4908a17150b49373b5f977685b2e180a04f6f/examples/server/server.cpp#L424

Probably a regression after the parallel decoding support to `server` was merged

**Comment by ReactorScram** [comment]
I could check it out as long as I'm here. I found the directions to run the CI locally, is there anything else I should do to test a patch?

**Comment by KerfuffleV2** [comment]
> Probably a regression after the parallel decoding support to `server` was merged

I don't think there's currently any sane way to handle this in `server` (or just any case where you're doing parallel generation and you want different sequences to use a different seed).

The problem is the RNG is stored in the context and you can't save/restore its state, you can only set the seed. I think some backend stuff would need to change before this could actually be fixed. Like putting the RNG into the sampler state stuff, then each sequence could have its own RNG initialized to different seeds.

**Comment by ReactorScram** [comment]
We can split this off if my issue on CPU is different, but on bbecf3f4 it seems to work with cache_prompt false? Whatever commit I was using before didn't work, I think.

![image](https://github.com/ggerganov/llama.cpp/assets/13400041/20056ad7-ba12-4cfe-bdf8-d902dc0a1dc1)

Is this intended behavior for cache_prompt? If not, is there a place I should add a regression test? I have a use case where I just need the LLM to be a classifier and I don't want it to be creative at all.

**Comment by KerfuffleV2** [comment]
Hmm, we might be talking about slightly different things. It looks like you're using temperature 0 so RNG stuff shouldn't affect you at all. The problem I was talking about would apply in the case when the RNG would have an effect (so temperature > 0) and also parallel sequences that should each use a different seed.

So basically I think you can pretty much disregard both my comment and GG's about the RNG seed thing. It should not make a difference in your particular case.

**Comment by ReactorScram** [comment]
Okay. I just happened to find my CPU inference issue by searching the issue tracker for "determinism" so I'm not sure if I should start a new issue or what, mine isn't CUDA-related.

**Comment by lmg-anon** [comment]
@ReactorScram I can confirm there's something wrong with `cache_prompt: true`.

**Comment by ReactorScram** [comment]
Okay cool. I don't know exactly how it's supposed to work but I assume "cache" means it should yield the same output but faster, right? I did notice the 2nd inference for the same prompt is usually quicker. But it doesn't feed in previous prompts, it just caches the state in case 2 prompts share a prefix?

**Comment by ReactorScram** [comment]
@ggerganov I tried un-commenting that line, but it doesn't seem to compile because the seed can only be set on the llama_context, which is server-wide, and the requests come to a specific slot, and the slot only has a sampling context, for which I cannot set the seed.

I noticed that even if I disable caching, the seed isn't used - At temperature 0.0, the output never changes even if the seed changes. At 0.7 the output always changes, even if the seed doesn't.

![image](https://github.com/ggerganov/llama.cpp/assets/13400041/782e99e7-3819-4f75-9a37-3944793a80a1)

I set up Doxygen in my fork so anything I learn as a new reader / contributor will have a place to be added. I'll make a PR for that if it's fine. Here's my local git log:

```
1650511 (HEAD -> master, me/master) server : extract assign_work_to_slot to reduce max indentation and function length
3fbfdec common : set up Doxygen comments in a few places
06de72f server : fix uninitialized members
a46349a docs : add Doxygen config
5d231da ci : document free disk space needed
b691e61 gitignore : build-info.h
```

**Comment by KerfuffleV2** [comment]
> I tried un-commenting that line, but it doesn't seem to compile

Yes, it got disabled when the sampling context stuff got added. The reason presumably was because it wasn't a simple fix.

But like I mentioned, that shouldn't really have anything to do with your issue because you're using temperature 0. If the issue only occurs when you have prompt caching turned on, then if you want to try to fix the issue you should probably look at the differences between when prompt caching is disabled vs enabled.

**Comment by d-kleine** [comment]
I confirm this behavior too. When disabling the KV cache, the output is deterministic. When keeping it enabled, the first output is different from the following ones (second, third, etc.), even when setting deterministic params (seed, temp set to 0, fixed ctx len, etc.). It seems like the cache is not updated with the same state as has been initialized.

**Comment by ivich123** [comment]
I've been investigating this issue and I believe I've identified the root cause.

**Environment:**
- GPU: NVIDIA RTX 5090 (compute 12.0)
- Driver: 590.48.01
- CUDA toolkit: 13.1 (compiled with arch 120)
- Model: TheBloke/Llama-2-7B-GGUF Q4_0
- llama.cpp build: 8071 (2ba9adc0)

**Bisection of ngl threshold:**

| ngl | Bug reproduced? |
|-----|-----------------|
| 0 (CPU) | No |
| 1 | No |
| 2 | No |
| 3 | Yes |
| 99 (all) | Yes |

**Root cause: batch size difference between first and subsequent evaluations**

I added debug logging to print the first 10 logits before sampling (at `common_sampler_sample` in `server-context.cpp:2746`). Comparing logits[0] at n_decoded=0 across requests:
```
Task 0   (1st request, fresh):        -15.448252
Task 51  (2nd request, cache reuse):  -15.436468
Task 102 (3rd request):               -15.436468
Task 153 (4th request):               -15.436468
```

After cache clear with a different prompt:
```
Task 257 (1st after clear):           -15.371330
Task 308 (2nd after clear):           -15.396111
Task 359 (3rd):                       -15.396111
Task 410 (4th):                       -15.396111
```

The first request always produces different logits. Subsequent requests are identical to each other but differ from the first.

**Why this happens:**

The server logs show the mechanism:

- **First request:** `n_tokens = 0, memory_seq_rm [0, end)` → evaluates all 5 prompt tokens in one batch (`batch.n_tokens = 5`)
- **Second request:** `n_past was set to 4`, `memory_seq_rm [4, end)` → reuses 4 cached tokens, evaluates only 1 token (`batch.n_tokens = 1`)

The server always re-evaluates at least 1 token because it needs the logits of the last prompt token to perform sampling. The KV cache stores Keys and Values but not logits, so the server discards the last cached token and re-evaluates it.

CUDA Flash Attention is not bit-identical when computing attention with different batch sizes (5 tokens at once vs 1 token reading K/V from cache). The floating point operation order differs due to parallelism, producing small numerical differences (~0.01-0.07) that accumulate across 32 transformer layers and are enough to change which token wins during sampling.

Additionally, the re-evaluated token produces different K/V values in the cache, so even if the first sampled token were corrected, subsequent tokens would still diverge.

This is consistent with all observations in this thread:
- Not observed on CPU (CPU attention is deterministic across batch sizes)
- Not observed with ngl 0-2 (not enough layers on GPU to accumulate the difference)
- Disabling `cache_prompt` fixes it (forces full re-evaluation every time)
- The first response matches what `llama-cli` returns (always evaluates from scratch)

**Proposed fix and open questions**

I have a working fix that forces `n_past = 0` when the user explicitly sets a seed (`seed != LLAMA_DEFAULT_SEED`), ensuring the prompt is always evaluated with the same batch size:
```cpp
// force full re-evaluation when seed is fixed to ensure determinism (fix #2838)
if (slot.task->params.sampling.seed != LLAMA_DEFAULT_SEED) {
    n_past = 0;
}
```

This works, but It's worth discussing whether this is the right approach before opening a PR:

1. **Is this redundant with `cache_prompt: false`?** Users can already disable prompt caching per request, which achieves the same effect. However, the current behavior is arguably a bug — users expect same seed + same prompt = same output without needing to know about internal caching details.

2. **Performance impact for long prompts with fixed seed.** This disables prompt cache reuse entirely when a seed is set. For short prompts this is negligible, but for long prompts (thousands of tokens) it could be noticeable. An alternative would be to only force re-evaluation when the prompt fully matches the cache (the `[TAG_PROMPT_LOGITS]` path), but that doesn't cover the case where a partial cache hit produces a different batch size.

3. **Is this a server-level fix or should it be documented as expected behavior?** CUDA Flash Attention being non-deterministic across batch sizes is arguably not a bug in llama.cpp — it's a fundamental property of parallel floating point. The question is whether the server should work around it transparently.

Happy to open a PR with any approach the maintainers prefer.

**Test script to reproduce:**
```bash
#!/bin/bash
NGL=${1:-99}
MODEL="$HOME/.cache/llama.cpp/TheBloke_Llama-2-7B-GGUF_llama-2-7b.Q4_0.gguf"
SERVER="./build/bin/llama-server"
PORT=8080

pkill -f llama-server 2>/dev/null; sleep 1
$SERVER -m $MODEL -ngl $NGL --port $PORT > server.log 2>&1 &
SERVER_PID=$!; sleep 3

echo "=== Test ngl=$NGL ==="
for i in $(seq 5); do
    curl -s -X POST --url "http://localhost:$PORT/completion" \
        --data '{"prompt":"Some random words:","n_predict":50,"seed":1337}' | jq -r '.content'
done

kill $SERVER_PID 2>/dev/null
```
